source/extensions/common/wasm/context.cc

```diff
@@ -1284,6 +1285,17 @@ WasmResult Context::defineMetric(uint32_t metric_type, std::string_view name,
   if (metric_type > static_cast<uint32_t>(MetricType::Max)) {
     return WasmResult::BadArgument;
   }
+  // Register prometheus namespace automatically so that
+  // custom metrics defined by Wasm plugins won't have "envoy_" prefix
+  // in the prometheus formatted metrics.
+  const auto aname = toAbslStringView(name);
+  const auto stat_namespace = aname.substr(0, aname.find_first_of('_'));
+  wasm()->dispatcher().post(
+      // Register the whole name if we couldn't find "_" in it.
+      [ns = stat_namespace.empty() ? std::string(name) : std::string(stat_namespace)]() -> void {
+        Server::PrometheusStatsFormatter::registerPrometheusNamespace(ns);
+      });
```

Comment:
I don't know enough about the current stats to know for sure, but isn't this too greedy? e.g. if someone defines `cluster_foo` metric in Wasm, wouldn't it automatically remove `envoy_` prefix from all "native" metrics as well?

Reply:
yeah that's true.. maybe better have something like "is_native" flag in Metric class, and prefix "envoy_" based on it, rather than using registerPrometheusNamespace?

Reply:
Yeah, that sounds good to me... although, it seems that such solution was discussed in https://github.com/envoyproxy/envoy/pull/11808, but the prefix list was added instead. cc @lizan @mandarjog @ggreenway @jmarantz

Reply:
I'm not familiar with wasm APIs at all, but could you require the plugin to register it's namespace, and clearly document that it should be specific to this plugin and not overlap with any existing native stat names, and that all plugin stats should be under that namespace? There's still the possibility of a plugin incorrectly registering for an incorrect prefix, but there's at least documentation about how to behave correctly to avoid such issues.

Reply:
I guess my point is why do we have those "namespaces" in the first place, when in reality it's only about separating native Envoy metrics that append `envoy_` prefix from those that don't?

Reply:
That's a good point. I think we could add another flag to the metric for this, don't set it by default, and have wasm always set it. @jmarantz any concerns with adding a new flag to metrics? I took a quick look, I think we have 16 bits of flags, and only use 3, so it shouldn't change the size of a Metric.

Reply:
>  @jmarantz any concerns with adding a new flag to metrics? I took a quick look, I think we have 16 bits of flags, and only use 3, so it shouldn't change the size of a Metric.

kindly ping @jmarantz, do you have any concerns about this?

Reply:
> Sorry I'm on vacation but I noticed this.

Oh sorry to interrupt you, and thanks for taking your time.

>  If this is really specific to wasm then it seems unclean to put a bit in the general stats structure to handle a special case.

This bit, say "is_custom_metric" flag, will be only set by Wasm *for now* since afaik Wasn extension is the only extension which defines non-Envoy-native metrics dynamically. But in the future, maybe other extensions would have that ability to define custom metrics.

Reply:
ping @jmarantz , wdyt ^^? Do you think the purpose is not worth a bit?

Reply:
@mathetake I think that we should do it, using flags to differentiate between Envoy's internal and Wasm/Lua/whatever metrics is much better UX than the need to register and use Prometheus namespaces.

Reply:
I'm fine with this.

Reply:
OK thank you all, I will rework this PR and add a "is_custom_metric" flag in Metric class to differentiate Envoy native metrics from others.